orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.h

```diff
@@ -0,0 +1,108 @@
+  std::unordered_map<size_t, int> output_port_reuse_recompute_with_compromise_count;
+
+  // Frequency Column
+  int freq = 0;
+};
+
+/**
+ * @brief Iterate the graph and find all possible memory optimization opportunities for related nodes.
+ *
+ * @param graph  The graph to iterate.
+ * @param probe_level The level to control allowed operations during recomputable subgraph detecting.
+ * @param logger Logger.
+ * @param node_index_to_its_order_in_topological_sort_map  The mapping of node index to its order in topological sort.
+ * @param yield_op_order_in_topological_sort The order of the boundary op in the topological sort.
+ * @param candidate_output_args_map  A map from node to its candidate activations, which are consumed by both fw and
+ * @param mem_opt_stats  A store to maintain all found optimization plans for related nodes.
+ * @return Status
+ */
+Status FindORTModuleMemoryOpportunity(const Graph& graph,
+                                      const ProbeLevel probe_level,
+                                      const logging::Logger& logger,
+                                      InlinedHashMap<NodeIndex, ptrdiff_t>&
+                                          node_index_to_its_order_in_topological_sort_map,
+                                      ptrdiff_t& yield_op_order_in_topological_sort,
+                                      InlinedHashMap<const Node*, InlinedVector<size_t>>& candidate_output_args_map,
+                                      MemoryOptimizationPlanner& mem_opt_stats);
+
+/**
+ * @brief From the optimization plans, generate the memory optimization statistics table containing many MemoryRecords,
+ * each represents one node cluster id.
+ *
+ * @param memory_opt_planner The optimization planner to get optimization plans.
+ * @param generated_records Returns the generated memory optimization statistics table.
+ * @param node_to_apply_contexts_map Optional. If provided, we will append the optimization applying information
+ * (for example, how many are actually applied) to each MemoryRecord.
+ */
+void GetMemoryRecordsGroupedByNodeClusterId(const MemoryOptimizationPlanner& memory_opt_planner,
+                                            std::vector<std::pair<std::string, MemoryRecord>>& generated_records,
+                                            const NodeToClusterApplyContextMap&
+                                                node_to_apply_contexts_map);
+
+/**
+ * @brief Serialize the memory optimization statistics table to a string.
+ *
+ * @param records_grouped_by_node_cluster_id The memory optimization statistics table.
+ * @param user_config Optional. If provided, we will append the user configuration to the serialized string.
+ * @return std::string
+ */
+std::string SerializeMemoryRecords(const std::vector<std::pair<std::string, MemoryRecord>>&
+                                       records_grouped_by_node_cluster_id,
+                                   const std::string user_config = "");
```

Comment:
>const std::string user_config  [](http://example.com/codeflow?start=35&length=30)

Any reason to use by value string arg?

Reply:
Used std::string_view, and make it an required parameter now.